// @ngneat/transloco-preload-langs/esm2020/lib/preload-langs.service.mjs
import { Inject, Injectable, InjectionToken } from '@angular/core';
import { TranslocoService } from '@ngneat/transloco';
import { tap } from 'rxjs/operators';
import { forkJoin } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@ngneat/transloco";
export const PRELOAD_LANGUAGES = new InjectionToken('Languages to be preloaded');
export class TranslocoPreloadLangsService {
    constructor(service, langs) {
        this.subscription = null;
        if (!langs)
            return;
        this.idleCallbackId = window.requestIdleCallback(() => {
            const preloads = langs.map((currentLangOrScope) => {
                const lang = service._completeScopeWithLang(currentLangOrScope);
                return service.load(lang).pipe(tap(() => {
                    if (service.config.prodMode === false) {
                        console.log(`%c 👁 Preloaded ${lang}`, 'background: #fff; color: #607D8B;');
                    }
                }));
            });
            this.subscription = forkJoin(preloads).subscribe();
        });
    }
    ngOnDestroy() {
        if (this.idleCallbackId !== undefined) {
            window.cancelIdleCallback(this.idleCallbackId);
        }
        this.subscription?.unsubscribe();
        // Caretaker note: it's important to clean up references to subscriptions since they save the `next`
        // callback within its `destination` property, preventing classes from being GC'd.
        this.subscription = null;
    }
}
TranslocoPreloadLangsService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsService, deps: [{ token: i1.TranslocoService }, { token: PRELOAD_LANGUAGES }], target: i0.ɵɵFactoryTarget.Injectable });
TranslocoPreloadLangsService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.TranslocoService }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [PRELOAD_LANGUAGES]
                }] }]; } });
