// @ngneat/transloco-preload-langs/esm2020/lib/preload-langs.module.mjs
import { APP_INITIALIZER, NgModule } from '@angular/core';
import { PRELOAD_LANGUAGES, TranslocoPreloadLangsService, } from './preload-langs.service';
import * as i0 from "@angular/core";
window.requestIdleCallback =
    window.requestIdleCallback ||
        function (cb) {
            const start = Date.now();
            return setTimeout(function () {
                cb({
                    didTimeout: false,
                    timeRemaining: function () {
                        return Math.max(0, 50 - (Date.now() - start));
                    },
                });
            }, 1);
        };
window.cancelIdleCallback =
    window.cancelIdleCallback ||
        function (id) {
            clearTimeout(id);
        };
// eslint-disable-next-line @typescript-eslint/no-empty-function
function noop() { }
export class TranslocoPreloadLangsModule {
    static forRoot(langs) {
        return {
            ngModule: TranslocoPreloadLangsModule,
            providers: [
                TranslocoPreloadLangsService,
                { provide: PRELOAD_LANGUAGES, useValue: langs },
                {
                    provide: APP_INITIALIZER,
                    useFactory: () => noop,
                    multi: true,
                    deps: [TranslocoPreloadLangsService]
                },
            ],
        };
    }
}
TranslocoPreloadLangsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TranslocoPreloadLangsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsModule });
TranslocoPreloadLangsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsModule, decorators: [{
            type: NgModule
        }] });
