// @ngneat/transloco-preload-langs/fesm2020/ngneat-transloco-preload-langs.mjs
import * as i0 from '@angular/core';
import { InjectionToken, Injectable, Inject, APP_INITIALIZER, NgModule } from '@angular/core';
import * as i1 from '@ngneat/transloco';
import { tap } from 'rxjs/operators';
import { forkJoin } from 'rxjs';

const PRELOAD_LANGUAGES = new InjectionToken('Languages to be preloaded');
class TranslocoPreloadLangsService {
    constructor(service, langs) {
        this.subscription = null;
        if (!langs)
            return;
        this.idleCallbackId = window.requestIdleCallback(() => {
            const preloads = langs.map((currentLangOrScope) => {
                const lang = service._completeScopeWithLang(currentLangOrScope);
                return service.load(lang).pipe(tap(() => {
                    if (service.config.prodMode === false) {
                        console.log(`%c ðŸ‘ Preloaded ${lang}`, 'background: #fff; color: #607D8B;');
                    }
                }));
            });
            this.subscription = forkJoin(preloads).subscribe();
        });
    }
    ngOnDestroy() {
        if (this.idleCallbackId !== undefined) {
            window.cancelIdleCallback(this.idleCallbackId);
        }
        this.subscription?.unsubscribe();
        // Caretaker note: it's important to clean up references to subscriptions since they save the `next`
        // callback within its `destination` property, preventing classes from being GC'd.
        this.subscription = null;
    }
}
TranslocoPreloadLangsService.Éµfac = i0.ÉµÉµngDeclareFactory({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsService, deps: [{ token: i1.TranslocoService }, { token: PRELOAD_LANGUAGES }], target: i0.ÉµÉµFactoryTarget.Injectable });
TranslocoPreloadLangsService.Éµprov = i0.ÉµÉµngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsService });
i0.ÉµÉµngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.TranslocoService }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [PRELOAD_LANGUAGES]
                }] }]; } });

window.requestIdleCallback =
    window.requestIdleCallback ||
        function (cb) {
            const start = Date.now();
            return setTimeout(function () {
                cb({
                    didTimeout: false,
                    timeRemaining: function () {
                        return Math.max(0, 50 - (Date.now() - start));
                    },
                });
            }, 1);
        };
window.cancelIdleCallback =
    window.cancelIdleCallback ||
        function (id) {
            clearTimeout(id);
        };
// eslint-disable-next-line @typescript-eslint/no-empty-function
function noop() { }
class TranslocoPreloadLangsModule {
    static forRoot(langs) {
        return {
            ngModule: TranslocoPreloadLangsModule,
            providers: [
                TranslocoPreloadLangsService,
                { provide: PRELOAD_LANGUAGES, useValue: langs },
                {
                    provide: APP_INITIALIZER,
                    useFactory: () => noop,
                    multi: true,
                    deps: [TranslocoPreloadLangsService]
                },
            ],
        };
    }
}
TranslocoPreloadLangsModule.Éµfac = i0.ÉµÉµngDeclareFactory({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsModule, deps: [], target: i0.ÉµÉµFactoryTarget.NgModule });
TranslocoPreloadLangsModule.Éµmod = i0.ÉµÉµngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsModule });
TranslocoPreloadLangsModule.Éµinj = i0.ÉµÉµngDeclareInjector({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsModule });
i0.ÉµÉµngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoPreloadLangsModule, decorators: [{
            type: NgModule
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TranslocoPreloadLangsModule };
