// @ngneat/transloco-locale/esm2020/lib/pipes/transloco-decimal.pipe.mjs
import { Pipe, ChangeDetectorRef, Inject, } from '@angular/core';
import { isNil } from '@ngneat/transloco';
import { getDefaultOptions } from '../shared';
import { LOCALE_CONFIG } from '../transloco-locale.config';
import { TranslocoLocaleService } from '../transloco-locale.service';
import { TranslocoLocalePipe } from './transloco-locale.pipe';
import * as i0 from "@angular/core";
import * as i1 from "../transloco-locale.service";
export class TranslocoDecimalPipe extends TranslocoLocalePipe {
    constructor(translocoLocaleService, cdr, localeConfig) {
        super(translocoLocaleService, cdr);
        this.translocoLocaleService = translocoLocaleService;
        this.cdr = cdr;
        this.localeConfig = localeConfig;
    }
    /**
     * Transform a given number into the locale's currency format.
     *
     * @example
     *
     * 1234567890 | translocoDecimal: {} : en-US // 1,234,567,890
     * 1234567890 | translocoDecimal: {useGrouping: false}: en-US // 1234567890
     *
     */
    transform(value, numberFormatOptions = {}, locale) {
        if (isNil(value))
            return '';
        locale = this.getLocale(locale);
        const options = {
            ...getDefaultOptions(locale, 'decimal', this.localeConfig),
            ...numberFormatOptions,
        };
        return this.translocoLocaleService.localizeNumber(value, 'decimal', locale, options);
    }
    ngOnDestroy() {
        super.onDestroy();
    }
}
TranslocoDecimalPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoDecimalPipe, deps: [{ token: i1.TranslocoLocaleService }, { token: i0.ChangeDetectorRef }, { token: LOCALE_CONFIG }], target: i0.ɵɵFactoryTarget.Pipe });
TranslocoDecimalPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoDecimalPipe, name: "translocoDecimal", pure: false });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoDecimalPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'translocoDecimal',
                    pure: false,
                }]
        }], ctorParameters: function () { return [{ type: i1.TranslocoLocaleService }, { type: i0.ChangeDetectorRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [LOCALE_CONFIG]
                }] }]; } });
