// @ngneat/transloco-locale/esm2020/lib/pipes/transloco-date.pipe.mjs
import { Pipe, ChangeDetectorRef, Inject, } from '@angular/core';
import { isNil } from '@ngneat/transloco';
import { getDefaultOptions } from '../shared';
import { LOCALE_CONFIG } from '../transloco-locale.config';
import { TranslocoLocaleService } from '../transloco-locale.service';
import { TranslocoLocalePipe } from './transloco-locale.pipe';
import * as i0 from "@angular/core";
import * as i1 from "../transloco-locale.service";
export class TranslocoDatePipe extends TranslocoLocalePipe {
    constructor(translocoLocaleService, cdr, localeConfig) {
        super(translocoLocaleService, cdr);
        this.translocoLocaleService = translocoLocaleService;
        this.cdr = cdr;
        this.localeConfig = localeConfig;
    }
    /**
     * Transform a date into the locale's date format.
     *
     * The date expression: a `Date` object,  a number
     * (milliseconds since UTC epoch), or an ISO string (https://www.w3.org/TR/NOTE-datetime).
     *
     * @example
     *
     * date | translocoDate: {} : en-US // 9/10/2019
     * date | translocoDate: { dateStyle: 'medium', timeStyle: 'medium' } : en-US // Sep 10, 2019, 10:46:12 PM
     * date | translocoDate: { timeZone: 'UTC', timeStyle: 'full' } : en-US // 7:40:32 PM Coordinated Universal Time
     * 1 | translocoDate: { dateStyle: 'medium' } // Jan 1, 1970
     * '2019-02-08' | translocoDate: { dateStyle: 'medium' } // Feb 8, 2019
     */
    transform(date, options = {}, locale) {
        if (isNil(date))
            return '';
        locale = this.getLocale(locale);
        return this.translocoLocaleService.localizeDate(date, locale, {
            ...getDefaultOptions(locale, 'date', this.localeConfig),
            ...options,
        });
    }
    ngOnDestroy() {
        super.onDestroy();
    }
}
TranslocoDatePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoDatePipe, deps: [{ token: i1.TranslocoLocaleService }, { token: i0.ChangeDetectorRef }, { token: LOCALE_CONFIG }], target: i0.ɵɵFactoryTarget.Pipe });
TranslocoDatePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoDatePipe, name: "translocoDate", pure: false });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoDatePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'translocoDate',
                    pure: false,
                }]
        }], ctorParameters: function () { return [{ type: i1.TranslocoLocaleService }, { type: i0.ChangeDetectorRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [LOCALE_CONFIG]
                }] }]; } });
