// @ngneat/transloco-locale/esm2020/lib/pipes/transloco-currency.pipe.mjs
import { Pipe, ChangeDetectorRef, Inject, } from '@angular/core';
import { isNil } from '@ngneat/transloco';
import { getDefaultOptions } from '../shared';
import { LOCALE_CONFIG } from '../transloco-locale.config';
import { TranslocoLocaleService } from '../transloco-locale.service';
import { TranslocoLocalePipe } from './transloco-locale.pipe';
import * as i0 from "@angular/core";
import * as i1 from "../transloco-locale.service";
export class TranslocoCurrencyPipe extends TranslocoLocalePipe {
    constructor(translocoLocaleService, cdr, localeConfig) {
        super(translocoLocaleService, cdr);
        this.translocoLocaleService = translocoLocaleService;
        this.cdr = cdr;
        this.localeConfig = localeConfig;
    }
    /**
     * Transform a given number into the locale's currency format.
     *
     * @example
     *
     * 1000000 | translocoCurrency: 'symbol' : {} : USD // $1,000,000.00
     * 1000000 | translocoCurrency: 'name' : {} : USD // 1,000,000.00 US dollars
     * 1000000 | translocoCurrency: 'symbol' : {minimumFractionDigits: 0 } : USD // $1,000,000
     *
     */
    transform(value, display = 'symbol', numberFormatOptions = {}, currencyCode, locale) {
        if (isNil(value))
            return '';
        locale = this.getLocale(locale);
        const options = {
            ...getDefaultOptions(locale, 'currency', this.localeConfig),
            ...numberFormatOptions,
            currencyDisplay: display,
            currency: currencyCode || this.translocoLocaleService._resolveCurrencyCode(),
        };
        return this.translocoLocaleService.localizeNumber(value, 'currency', locale, options);
    }
    ngOnDestroy() {
        super.onDestroy();
    }
}
TranslocoCurrencyPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoCurrencyPipe, deps: [{ token: i1.TranslocoLocaleService }, { token: i0.ChangeDetectorRef }, { token: LOCALE_CONFIG }], target: i0.ɵɵFactoryTarget.Pipe });
TranslocoCurrencyPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoCurrencyPipe, name: "translocoCurrency", pure: false });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoCurrencyPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'translocoCurrency',
                    pure: false,
                }]
        }], ctorParameters: function () { return [{ type: i1.TranslocoLocaleService }, { type: i0.ChangeDetectorRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [LOCALE_CONFIG]
                }] }]; } });
