// ngx-isr/esm2020/lib/http-errors.interceptor.mjs
import { Injectable } from '@angular/core';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { catchError, throwError } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "./ngx-isr.service";
export class HttpErrorsInterceptor {
    constructor(ngxIsrService) {
        this.ngxIsrService = ngxIsrService;
    }
    intercept(request, next) {
        return next.handle(request).pipe(catchError(err => {
            this.ngxIsrService.addError(err);
            return throwError(() => err);
        }));
    }
}
HttpErrorsInterceptor.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: HttpErrorsInterceptor, deps: [{ token: i1.NgxIsrService }], target: i0.ɵɵFactoryTarget.Injectable });
HttpErrorsInterceptor.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: HttpErrorsInterceptor });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: HttpErrorsInterceptor, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.NgxIsrService }]; } });
export const HTTP_ERROR_PROVIDER_ISR = {
    provide: HTTP_INTERCEPTORS,
    useClass: HttpErrorsInterceptor,
    multi: true,
};
