// ngx-isr/esm2020/lib/ngx-isr.service.mjs
import { Inject, Injectable, PLATFORM_ID } from '@angular/core';
import { DOCUMENT, isPlatformServer } from '@angular/common';
import { ChildActivationEnd } from '@angular/router';
import { filter, map, take } from 'rxjs/operators';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
const initialState = {
    revalidate: null,
    errors: [],
    extra: {}
};
export class NgxIsrService {
    constructor(platformId, doc, router) {
        this.platformId = platformId;
        this.doc = doc;
        this.router = router;
        this.state = new BehaviorSubject(initialState);
        this.setRevalidate = (revalidate) => {
            this.state.next({ ...this.getState(), revalidate });
        };
        if (isPlatformServer(this.platformId)) {
            this.activate();
        }
    }
    getState() {
        return this.state.getValue();
    }
    getExtra() {
        return this.state.getValue().extra;
    }
    activate() {
        this.router.events
            .pipe(filter((e) => e instanceof ChildActivationEnd), map((event) => {
            let snapshot = event.snapshot;
            while (snapshot.firstChild !== null) {
                snapshot = snapshot.firstChild;
            }
            return snapshot.data;
        }), take(1))
            .subscribe((data) => {
            if (data?.['revalidate'] !== undefined) {
                this.setRevalidate(data['revalidate']);
            }
        });
    }
    addError(err) {
        const currentErrors = this.getState().errors;
        this.state.next({ ...this.getState(), errors: [...currentErrors, err] });
    }
    addExtra(extra = {}) {
        this.state.next({ ...this.getState(), extra: { ...this.getExtra(), ...extra } });
    }
}
NgxIsrService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxIsrService, deps: [{ token: PLATFORM_ID }, { token: DOCUMENT }, { token: i1.Router }], target: i0.ɵɵFactoryTarget.Injectable });
NgxIsrService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxIsrService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxIsrService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: Object, decorators: [{
                    type: Inject,
                    args: [PLATFORM_ID]
                }] }, { type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }, { type: i1.Router }]; } });
