// ngx-isr/esm2020/lib/ngx-isr.module.mjs
import { NgModule } from '@angular/core';
import { NgxIsrService } from './ngx-isr.service';
import { HTTP_ERROR_PROVIDER_ISR } from './http-errors.interceptor';
import { BEFORE_APP_SERIALIZED } from '@angular/platform-server';
import { addIsrDataBeforeSerialized } from './utils/add-isr-data-before-serialized';
import { DOCUMENT } from '@angular/common';
import * as i0 from "@angular/core";
import * as i1 from "./ngx-isr.service";
export class NgxIsrModule {
    constructor(isrService) {
        this.isrService = isrService;
    }
    static forRoot() {
        return {
            ngModule: NgxIsrModule,
            providers: [
                NgxIsrService,
                HTTP_ERROR_PROVIDER_ISR,
                {
                    provide: BEFORE_APP_SERIALIZED,
                    useFactory: addIsrDataBeforeSerialized,
                    multi: true,
                    deps: [NgxIsrService, DOCUMENT]
                },
            ]
        };
    }
}
NgxIsrModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxIsrModule, deps: [{ token: i1.NgxIsrService }], target: i0.ɵɵFactoryTarget.NgModule });
NgxIsrModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.4", ngImport: i0, type: NgxIsrModule });
NgxIsrModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxIsrModule, providers: [NgxIsrService] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.4", ngImport: i0, type: NgxIsrModule, decorators: [{
            type: NgModule,
            args: [{ providers: [NgxIsrService] }]
        }], ctorParameters: function () { return [{ type: i1.NgxIsrService }]; } });
