// @ngneat/transloco-locale/esm2020/lib/transloco-locale.service.mjs
import { Injectable, Inject } from '@angular/core';
import { TranslocoService } from '@ngneat/transloco';
import { BehaviorSubject } from 'rxjs';
import { map, distinctUntilChanged, filter } from 'rxjs/operators';
import { isLocaleFormat, toDate } from './helpers';
import { getDefaultOptions } from './shared';
import { LOCALE_LANG_MAPPING, LOCALE_DEFAULT_LOCALE, LOCALE_CONFIG, LOCALE_CURRENCY_MAPPING, LOCALE_DEFAULT_CURRENCY, } from './transloco-locale.config';
import { TRANSLOCO_DATE_TRANSFORMER, TRANSLOCO_NUMBER_TRANSFORMER, } from './transloco-locale.transformers';
import * as i0 from "@angular/core";
import * as i1 from "@ngneat/transloco";
export class TranslocoLocaleService {
    constructor(translocoService, langLocaleMapping, defaultLocale, defaultCurrency, localeConfig, localeCurrencyMapping, numberTransformer, dateTransformer) {
        this.translocoService = translocoService;
        this.langLocaleMapping = langLocaleMapping;
        this.defaultLocale = defaultLocale;
        this.defaultCurrency = defaultCurrency;
        this.localeConfig = localeConfig;
        this.localeCurrencyMapping = localeCurrencyMapping;
        this.numberTransformer = numberTransformer;
        this.dateTransformer = dateTransformer;
        this._locale = this.defaultLocale || this.toLocale(this.translocoService.getActiveLang());
        this.locale = new BehaviorSubject(this._locale);
        this.subscription = this.translocoService.langChanges$
            .pipe(map(this.toLocale.bind(this)), filter((locale) => !!locale))
            .subscribe({
            next: (locale) => this.setLocale(locale),
        });
        this.localeChanges$ = this.locale
            .asObservable()
            .pipe(distinctUntilChanged());
    }
    getLocale() {
        return this._locale;
    }
    setLocale(locale) {
        if (!isLocaleFormat(locale)) {
            console.error(`${locale} isn't a valid locale format`);
            return;
        }
        this.locale.next(locale);
        this._locale = locale;
    }
    /**
     * Get the currency symbol for the currently set locale.
     */
    getCurrencySymbol(locale = this.getLocale()) {
        const currency = this.localeCurrencyMapping[locale];
        const numberFormat = new Intl.NumberFormat(locale, {
            style: 'currency',
            currencyDisplay: 'symbol',
            currency: currency,
            minimumFractionDigits: 0,
            maximumFractionDigits: 0,
        });
        const pivot = 0;
        return numberFormat
            .format(pivot)
            .split(pivot.toString())
            .map((element) => element.trim())
            .find((element) => !!element);
    }
    /**
     * Transform a date into the locale's date format.
     *
     * The date expression: a `Date` object, a number
     * (milliseconds since UTC epoch), or an ISO string (https://www.w3.org/TR/NOTE-datetime).
     *
     * @example
     *
     * localizeDate(new Date(2019, 9, 7, 12, 0, 0)) // 10/7/2019
     * localizeDate(date, 'en-US', { dateStyle: 'medium', timeStyle: 'medium' }) // Sep 10, 2019, 10:46:12 PM
     * localizeDate(date) 'en-US', { timeZone: 'UTC', timeStyle: 'full' } // 7:40:32 PM Coordinated Universal Time
     * localizeDate(1, 'en-US', { dateStyle: 'medium' }) // Jan 1, 1970
     * localizeDate('2019-02-08', 'en-US', { dateStyle: 'medium' }) // Feb 8, 2019
     */
    localizeDate(date, locale = this.getLocale(), options = {}) {
        options = options
            ? options
            : getDefaultOptions(locale, 'date', this.localeConfig);
        return this.dateTransformer.transform(toDate(date), locale, options);
    }
    /**
     * Transform a number into the locale's number format according to the number type.
     *
     * localizeNumber(1234567890, 'decimal') // 1,234,567,890
     * localizeNumber(0.5, 'percent') // 50%
     * localizeNumber(1000, 'currency') // $1,000.00
     */
    localizeNumber(value, type, locale = this.getLocale(), options) {
        let resolved = options ?? getDefaultOptions(locale, type, this.localeConfig);
        if (type === 'currency') {
            resolved = {
                ...resolved,
                currency: resolved.currency || this._resolveCurrencyCode(locale),
            };
        }
        return this.numberTransformer.transform(value, type, locale, resolved);
    }
    /**
     * @internal
     */
    _resolveCurrencyCode(locale = this.getLocale()) {
        return this.localeCurrencyMapping[locale] || this.defaultCurrency;
    }
    ngOnDestroy() {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        this.subscription.unsubscribe();
        // Caretaker note: it's important to clean up references to subscriptions since they save the `next`
        // callback within its `destination` property, preventing classes from being GC'd.
        this.subscription = null;
    }
    toLocale(val) {
        if (this.langLocaleMapping[val]) {
            return this.langLocaleMapping[val];
        }
        if (isLocaleFormat(val)) {
            return val;
        }
        return '';
    }
}
TranslocoLocaleService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoLocaleService, deps: [{ token: i1.TranslocoService }, { token: LOCALE_LANG_MAPPING }, { token: LOCALE_DEFAULT_LOCALE }, { token: LOCALE_DEFAULT_CURRENCY }, { token: LOCALE_CONFIG }, { token: LOCALE_CURRENCY_MAPPING }, { token: TRANSLOCO_NUMBER_TRANSFORMER }, { token: TRANSLOCO_DATE_TRANSFORMER }], target: i0.ɵɵFactoryTarget.Injectable });
TranslocoLocaleService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoLocaleService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoLocaleService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.TranslocoService }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [LOCALE_LANG_MAPPING]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [LOCALE_DEFAULT_LOCALE]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [LOCALE_DEFAULT_CURRENCY]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [LOCALE_CONFIG]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [LOCALE_CURRENCY_MAPPING]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TRANSLOCO_NUMBER_TRANSFORMER]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TRANSLOCO_DATE_TRANSFORMER]
                }] }]; } });
