// @ngneat/transloco-locale/esm2020/lib/transloco-locale.module.mjs
import { NgModule } from '@angular/core';
import { TranslocoCurrencyPipe } from './pipes/transloco-currency.pipe';
import { TranslocoDatePipe } from './pipes/transloco-date.pipe';
import { TranslocoDecimalPipe } from './pipes/transloco-decimal.pipe';
import { TranslocoPercentPipe } from './pipes/transloco-percent.pipe';
import { LOCALE_CURRENCY_MAPPING, LOCALE_LANG_MAPPING, defaultConfig, LOCALE_DEFAULT_LOCALE, LOCALE_CONFIG, LOCALE_DEFAULT_CURRENCY, } from './transloco-locale.config';
import { TRANSLOCO_DATE_TRANSFORMER, TRANSLOCO_NUMBER_TRANSFORMER, DefaultDateTransformer, DefaultNumberTransformer, } from './transloco-locale.transformers';
import * as i0 from "@angular/core";
export const pipes = [
    TranslocoCurrencyPipe,
    TranslocoDatePipe,
    TranslocoDecimalPipe,
    TranslocoPercentPipe,
];
export class TranslocoLocaleModule {
    static forRoot(config = {}) {
        return {
            ngModule: TranslocoLocaleModule,
            providers: [
                {
                    provide: LOCALE_LANG_MAPPING,
                    useValue: config.langToLocaleMapping || defaultConfig.langToLocaleMapping,
                },
                {
                    provide: LOCALE_CONFIG,
                    useValue: config.localeConfig || defaultConfig.localeConfig,
                },
                {
                    provide: LOCALE_CURRENCY_MAPPING,
                    useValue: config.localeToCurrencyMapping ||
                        defaultConfig.localeToCurrencyMapping,
                },
                {
                    provide: LOCALE_DEFAULT_LOCALE,
                    useValue: config.defaultLocale || defaultConfig.defaultLocale,
                },
                {
                    provide: LOCALE_DEFAULT_CURRENCY,
                    useValue: config.defaultCurrency || defaultConfig.defaultCurrency,
                },
                {
                    provide: TRANSLOCO_DATE_TRANSFORMER,
                    useClass: DefaultDateTransformer,
                },
                {
                    provide: TRANSLOCO_NUMBER_TRANSFORMER,
                    useClass: DefaultNumberTransformer,
                },
            ],
        };
    }
}
TranslocoLocaleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoLocaleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TranslocoLocaleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoLocaleModule, declarations: [TranslocoCurrencyPipe,
        TranslocoDatePipe,
        TranslocoDecimalPipe,
        TranslocoPercentPipe], exports: [TranslocoCurrencyPipe,
        TranslocoDatePipe,
        TranslocoDecimalPipe,
        TranslocoPercentPipe] });
TranslocoLocaleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoLocaleModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.0", ngImport: i0, type: TranslocoLocaleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: pipes,
                    exports: pipes,
                }]
        }] });
