// ngx-hover-preload/esm2015/lib/link.directive.js
import { Directive, Optional, } from '@angular/core';
import { RouterLink, RouterLinkWithHref, RouterPreloader } from '@angular/router';
import { RegistryService } from './registry.service';
export class LinkDirective {
    constructor(_loader, _registry, link, linkWithHref) {
        this._loader = _loader;
        this._registry = _registry;
        this._rl = link || linkWithHref;
    }
    prefetch() {
        requestIdleCallback(() => {
            this._registry.add(this._rl.urlTree);
            this._loader.preload().subscribe(() => void 0);
        });
    }
}
LinkDirective.decorators = [
    { type: Directive, args: [{
                selector: '[routerLink]',
                host: {
                    '(mouseenter)': 'prefetch()'
                }
            },] }
];
LinkDirective.ctorParameters = () => [
    { type: RouterPreloader },
    { type: RegistryService },
    { type: RouterLink, decorators: [{ type: Optional }] },
    { type: RouterLinkWithHref, decorators: [{ type: Optional }] }
];
const ɵ0 = function (cb) {
    const start = Date.now();
    return setTimeout(function () {
        cb({
            didTimeout: false,
            timeRemaining: function () {
                return Math.max(0, 50 - (Date.now() - start));
            }
        });
    }, 1);
}, ɵ1 = () => { };
const requestIdleCallback = typeof window !== 'undefined'
    ? window.requestIdleCallback || ɵ0
    : ɵ1;
export { ɵ0, ɵ1 };
