// ngx-hover-preload/esm2015/lib/registry.service.js
import { Injectable } from "@angular/core";
import { PRIMARY_OUTLET, Router } from "@angular/router";
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export class RegistryService {
    constructor(router) {
        this.router = router;
        // Using a map so we don't add an element
        // on each hover. This will generally reduce memory
        // usage and will not require cleanup.
        this._queue = new Set([]);
    }
    add(route) {
        this._queue.add(route);
    }
    shouldPrefetch(url) {
        const tree = this.router.parseUrl(url);
        return [...this._queue].some(containsTree.bind(null, tree));
    }
}
RegistryService.ɵprov = i0.ɵɵdefineInjectable({ factory: function RegistryService_Factory() { return new RegistryService(i0.ɵɵinject(i1.Router)); }, token: RegistryService, providedIn: "root" });
RegistryService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
RegistryService.ctorParameters = () => [
    { type: Router }
];
function containsQueryParams(container, containee) {
    // TODO: This does not handle array params correctly.
    return (Object.keys(containee).length <= Object.keys(container).length &&
        Object.keys(containee).every(key => containee[key] === container[key]));
}
function containsTree(containee, container) {
    return (containsQueryParams(container.queryParams, containee.queryParams) &&
        containsSegmentGroup(container.root, containee.root, containee.root.segments));
}
function containsSegmentGroup(container, containee, containeePaths) {
    if (container.segments.length > containeePaths.length) {
        const current = container.segments.slice(0, containeePaths.length);
        if (!equalPath(current, containeePaths))
            return false;
        if (containee.hasChildren())
            return false;
        return true;
    }
    else if (container.segments.length === containeePaths.length) {
        if (!equalPath(container.segments, containeePaths))
            return false;
        if (!containee.hasChildren())
            return true;
        for (const c in containee.children) {
            if (!container.children[c])
                break;
            if (containsSegmentGroup(container.children[c], containee.children[c], containee.children[c].segments))
                return true;
        }
        return false;
    }
    else {
        const current = containeePaths.slice(0, container.segments.length);
        const next = containeePaths.slice(container.segments.length);
        if (!equalPath(container.segments, current))
            return false;
        if (!container.children[PRIMARY_OUTLET])
            return false;
        return containsSegmentGroup(container.children[PRIMARY_OUTLET], containee, next);
    }
}
export function equalPath(as, bs) {
    if (as.length !== bs.length)
        return false;
    return as.every((a, i) => a.path === bs[i].path || a.path.startsWith(':') || bs[i].path.startsWith(':'));
}
