// @ngneat/transloco-messageformat/esm2020/lib/messageformat.module.mjs
import { NgModule } from '@angular/core';
import { TRANSLOCO_TRANSPILER } from '@ngneat/transloco';
import { MessageFormatTranspiler } from './messageformat.transpiler';
import { TRANSLOCO_MESSAGE_FORMAT_CONFIG, } from './messageformat.config';
import * as i0 from "@angular/core";
export class TranslocoMessageFormatModule {
    static forRoot(config) {
        return {
            ngModule: TranslocoMessageFormatModule,
            providers: [
                { provide: TRANSLOCO_MESSAGE_FORMAT_CONFIG, useValue: config },
                {
                    provide: TRANSLOCO_TRANSPILER,
                    useClass: MessageFormatTranspiler,
                },
            ],
        };
    }
}
TranslocoMessageFormatModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.10", ngImport: i0, type: TranslocoMessageFormatModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TranslocoMessageFormatModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.10", ngImport: i0, type: TranslocoMessageFormatModule });
TranslocoMessageFormatModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.10", ngImport: i0, type: TranslocoMessageFormatModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.10", ngImport: i0, type: TranslocoMessageFormatModule, decorators: [{
            type: NgModule
        }] });
