// @ngneat/transloco-messageformat/esm2020/lib/messageformat.transpiler.mjs
import { Inject, Injectable, Optional } from '@angular/core';
import { DefaultTranspiler, getValue, isObject, setValue, TRANSLOCO_CONFIG, } from '@ngneat/transloco';
import { TRANSLOCO_MESSAGE_FORMAT_CONFIG, } from './messageformat.config';
import { cachedFactory, defaultFactory, } from './messageformat.factory';
import * as i0 from "@angular/core";
export class MessageFormatTranspiler extends DefaultTranspiler {
    constructor(config, userConfig) {
        super(userConfig);
        const { locales, enableCache = true, ...messageConfig } = { locales: null, ...config };
        this.messageConfig = messageConfig;
        this.mfFactory = enableCache ? cachedFactory : defaultFactory;
        this.messageFormat = this.mfFactory(locales, messageConfig);
    }
    transpile(value, params = {}, translation, key) {
        if (!value) {
            return value;
        }
        if (isObject(value) && params) {
            Object.keys(params).forEach((p) => {
                const v = getValue(value, p);
                const getParams = getValue(params, p);
                const transpiled = super.transpile(v, getParams, translation, key);
                const message = this.messageFormat.compile(transpiled);
                value = setValue(value, p, message(params[p]));
            });
        }
        else if (!Array.isArray(value)) {
            const transpiled = super.transpile(value, params, translation, key);
            const message = this.messageFormat.compile(transpiled);
            return message(params);
        }
        return value;
    }
    onLangChanged(lang) {
        this.setLocale(lang);
    }
    setLocale(locale) {
        this.messageFormat = this.mfFactory(locale, this.messageConfig);
    }
}
MessageFormatTranspiler.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.10", ngImport: i0, type: MessageFormatTranspiler, deps: [{ token: TRANSLOCO_MESSAGE_FORMAT_CONFIG, optional: true }, { token: TRANSLOCO_CONFIG, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
MessageFormatTranspiler.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.10", ngImport: i0, type: MessageFormatTranspiler });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.10", ngImport: i0, type: MessageFormatTranspiler, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [TRANSLOCO_MESSAGE_FORMAT_CONFIG]
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [TRANSLOCO_CONFIG]
                }] }]; } });
